File: year2017/day01/solution.lua
local ffi = require('ffi')

ffi.cdef([[
int count_sum(const char* c, int len);
int count_sum_v2(const char* c, int len);
]])

local slib = ffi.load('solution')

---@param s string
---@return integer
local function count_sum(s)
    return slib.count_sum(s, #s)
end

---@param s string
---@return integer
local function count_sum_v2(s)
    return slib.count_sum_v2(s, #s)
end

local M = {}
M.count_sum = count_sum
M.count_sum_v2 = count_sum_v2
return M

File: year2017/day01/solution.c
int count_sum(const char* str, int len)
{
    int sum = 0;

    for (int i = 0; i < len; i++) {
        int j = i + 1;
        if (j >= len) {
            j = 0;
        }
        if (str[i] == str[j]) {
            sum += str[i] - '0';
        }
    }

    return sum;
}

int count_sum_v2(const char* str, int len)
{
    int sum = 0;
    int half = len / 2;

    for (int i = 0; i < half; i++) {
        if (str[i] == str[i + half]) {
            sum += str[i] - '0';
        }
    }

    return sum * 2;
}
